File: O/source_code_3.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main()
{
    int n;
    cout << "Enter the number of vertices: ";
    cin >> n;

    vector<vector<int>> graph(n, vector<int>(n));
    vector<int> degree(n, 0), color(n, 0), vertex(n);

    cout << "Enter adjacency matrix:\n";
    for (int i = 0; i < n; i++)
    {
        vertex[i] = i;
        for (int j = 0; j < n; j++)
        {
            cin >> graph[i][j];
            if (graph[i][j])
                degree[i]++;
        }
    }

    sort(vertex.begin(), vertex.end(), [&](int a, int b)
         { return degree[a] > degree[b]; });

    int currentColor = 1;
    for (int i = 0; i < n; i++)
    {
        if (color[vertex[i]] == 0)
        {
            color[vertex[i]] = currentColor;
            for (int j = i + 1; j < n; j++)
            {
                bool canColor = true;
                for (int k = 0; k < n; k++)
                {
                    if (graph[vertex[j]][k] && color[k] == currentColor)
                    {
                        canColor = false;
                        break;
                    }
                }
                if (canColor && color[vertex[j]] == 0)
                    color[vertex[j]] = currentColor;
            }
            currentColor++;
        }
    }

    cout << "\nVertex\tColor\n";
    for (int i = 0; i < n; i++)
    {
        cout << i << "\t" << color[i] << endl;
    }
    return 0;
}

File: O/source_code_3.py
def welch_powell_coloring(graph):
    n = len(graph)
    degree = [sum(row) for row in graph]
    vertex = list(range(n))

    # Sort vertices by degree descending
    vertex.sort(key=lambda x: -degree[x])

    color = [0] * n
    current_color = 1

    for i in range(n):
        if color[vertex[i]] == 0:
            color[vertex[i]] = current_color
            for j in range(i + 1, n):
                can_color = True
                for k in range(n):
                    if graph[vertex[j]][k] and color[k] == current_color:
                        print(graph[vertex[j]][k], color[k])
                        can_color = False
                        break
                if can_color and color[vertex[j]] == 0:
                    color[vertex[j]] = current_color
            current_color += 1

    return color


# Input
n = int(input("Enter the number of vertices: "))
graph = []
print("Enter adjacency matrix:")
for i in range(n):
    row = list(map(int, input().split()))
    graph.append(row)

# Run coloring algorithm
colors = welch_powell_coloring(graph)

# Output
print("\nVertex\tColor")
for i in range(n):
    print(f"{i}\t{colors[i]}")
